File: data_structures/filters/bloom_filter/bloomFilter.js
/*

A Bloom Filter is a space-efficient probabilistic data structure used to test whether an element
is a member of a set.

It uses multiple hash functions to set bits in a bit array, allowing quick membership checks.

However, it may produce false positives (indicating an element is in the set when it's not) but
never false negatives (indicating an element is not in the set when it is). This makes Bloom
Filters useful for applications where space is limited and some uncertainty is acceptable.

*/

class BloomFilter {
    constructor(size = 100) {
        this.size = size;
        this.bitArray = new Array(size).fill(0);
        this.hashFunctions = [
            (str) => this.hash(str, 17),
            (str) => this.hash(str, 31),
            (str) => this.hash(str, 37),
        ];
    }

    /**
     * A simple hashing function used to generate a value for the inputted string
     *
     * @param {string} str The string to hash based on the seed
     * @param {number} seed The seed to be used for the hashing function
     * @returns  {number} The hashed value for the string
     */
    hash(str, seed) {
        let hash = 0;
        for (let i = 0; i < str.length; i++) {
            hash = hash * seed + str.charCodeAt(i);
            hash = hash % this.size;
        }
        return hash;
    }

    /**
     * Adds a value to the probabilistic bloom filter data structure
     *
     * @param {string} str The string to hash and add to the bloom filter
     */
    add(str) {
        this.hashFunctions.forEach((fn) => {
            const index = fn(str);
            this.bitArray[index] = 1;
        });
    }

    /**
     * Checks if a string value exists in the bloom filter. This function returns a probabilistic
     * 'yes' (maybe it exists) and a firm 'no' (it definitely does not).
     *
     * @param {string} str A string to check for existence in the bloom filter
     * @returns {boolean} Returns a probabilistic 'yes' or a deterministic 'no' that the value is
     * in the filter
     */
    contains(str) {
        return this.hashFunctions.every((fn) => {
            const index = fn(str);
            return this.bitArray[index] === 1;
        });
    }
}

module.exports = BloomFilter;

File: data_structures/filters/bloom_filter/bloom_filter.py
"""
A Bloom Filter is a space-efficient probabilistic data structure used to test whether an element
is a member of a set.

It uses multiple hash functions to set bits in a bit array, allowing quick membership checks.

However, it may produce false positives (indicating an element is in the set when it's not) but
never false negatives (indicating an element is not in the set when it is). This makes Bloom
Filters useful for applications where space is limited and some uncertainty is acceptable.
"""


class BloomFilter:
    """
    This class implements a BloomFilter
    """

    def __init__(self, size=100):
        """
        Initializes the Bloom Filter with a specified size and sets up the bit array and hash
        functions.

        Parameters:
        size (int): The size of the bit array. Defaults to 100.
        """
        self.size = size
        self.bit_array = [0 for i in range(size)]
        self.hash_functions = [
            lambda a: self.hash(a, 17),
            lambda a: self.hash(a, 31),
            lambda a: self.hash(a, 37)
        ]

    def hash(self, string, seed):
        """
        Hashes a string using a specified seed and returns the hash value.

        Parameters:
        string (str): The input string to be hashed.
        seed (int): The seed value for the hash function.

        Returns:
        int: The hash value of the input string.
        """
        hash_value = 0
        for i, _ in enumerate(string):
            hash_value = (hash_value * seed) + ord(string[i])
            hash_value = hash_value % self.size

        return hash_value

    def add(self, item):
        """
        Adds an item to the Bloom Filter by setting the bits in the bit array using hash functions.

        Parameters:
        item (str): The item to be added to the Bloom Filter.
        """
        for fn in self.hash_functions:
            index = fn(item)
            self.bit_array[index] = 1

    def contains(self, item):
        """
        Checks if an item is possibly in the Bloom Filter.

        Parameters:
        item (str): The item to be checked.

        Returns:
        bool: True if the item is possibly in the Bloom Filter, False if it is definitely not.
        """
        for fn in self.hash_functions:
            index = fn(item)

            if self.bit_array[index] == 0:
                return False

        return True
